### PR TITLE
Prototype to create a separate sync module

### LINKED ISSUE #428: Seperate synchronization APIs from data access APIs
**Describe the issue to be researched**
The APIs for local resources and sync are quite tightly coupled at the moment. The `FhirEngine` interface definitely doesn't follow the single-responsibility principle.

**Describe the goal of the research**
Find out if there's a way to decouple the core storage APIs and sync APIs. 

**Describe the methodology**
Prototyping. Researching other similar codebases.

**Would you like to work on the issue?**
Yes :)

### PR DESCRIPTION
**Pull request by jingtang10**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #428 

**Description**
Separate the sync code into a new module. This is done via creating clear separation of concerns in the `FhirEngine` API.

```
suspend fun syncDownload(download: suspend (suspend (ResourceType) -> String?) -> List<Resource>)
suspend fun syncUpload(upload: (suspend (List<Resource>) -> Unit)?)
```

**Alternative(s) considered**
This is an improvement on the existing approach.

**Type**
Choose one: Feature

**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md)
- [x] I have read [How to Contribute](https://github.com/google/android-fhir/blob/master/docs/contributing.md)
- [x] I have read the [Developer's guide](https://github.com/google/android-fhir/wiki/Developer's-Guide)
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate )
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally
- [x] I have built and run the reference app(s) to verify my change fixes the issue and/or does not break the reference app(s)

### CODE REVIEW

File: core/src/main/java/com/google/android/fhir/impl/FhirEngineImpl.kt [hunk 1]
```diff
@@ -69,49 +53,21 @@ constructor(
...
-  }
-
   override suspend fun <R : Resource> search(search: Search): List<R> {
     return search.execute(database)
   }
 
-  private fun setupNextDownload(syncConfig: PeriodicSyncConfiguration) {
-    setupDownload(syncConfig = syncConfig, withInitialDelay = true)
+  override suspend fun syncDownload(download: suspend (suspend (ResourceType) -> String?) -> List<Resource>) {
+    val stuff = download(database::lastUpdate)
```
Comment (yigit) [inline core/src/main/java/com/google/android/fhir/impl/FhirEngineImpl.kt:61]:
```
interface SyncDownloadContext {
   suspend fun getLatestTimestamptFor(type:ResourceType):String?
}
```

```
override suspend fun syncDownload(downloader: suspend SyncDownloadContext.() -> List<Resource>)
```

or instead, we could just have a method that returns 

```
getTimestmaps(resourceTypes: List<ResourceType>): Map<ResourceType, Timeestamp>
```
```

File: sync/src/androidTest/java/com/google/android/fhir/sync/ExampleInstrumentedTest.kt [hunk 2]
```diff
@@ -0,0 +1,24 @@
+package com.google.android.fhir.sync
```
Comment (deepankarb) [inline sync/src/androidTest/java/com/google/android/fhir/sync/ExampleInstrumentedTest.kt:1]: Boilerplate.

File: sync/src/main/java/com/google/android/fhir/sync/Config.kt [hunk 3]
```diff
@@ -17,20 +17,44 @@
...
 import java.util.concurrent.TimeUnit
+import org.hl7.fhir.r4.model.ResourceType
+
+/**
+ * Class that holds what type of resources we need to synchronise and what are the parameters of
+ * that type. e.g. we only want to synchronise patients that live in United States
+ * `SyncData(ResourceType.Patient, mapOf("address-country" to "United States")`
+ */
+typealias ParamMap = Map<String, String>
+typealias SyncData = Map<ResourceType, ParamMap>
```
Comment (deepankarb) [inline sync/src/main/java/com/google/android/fhir/sync/Config.kt:30]: Can we improve on the naming `SyncData`. Off the cuff it sounds like data that needs be synced whereas this is more like Sync parameters by resource. Maybe `ResourceSyncParams`?

**Review by deepankarb — COMMENTED** [review]
Thanks Jing! Left a few comments.

**Comment by jingtang10** [comment]
closing this as this is being reworked in #482

**Pull request closed without merging**